server/src/main/java/com/google/fhir/proxy/HttpFhirClient.java

```diff
@@ -43,7 +43,38 @@ public abstract class HttpFhirClient {
 
   // The list of header names to keep in a response sent from the proxy; use lower case only.
   // Note we don't copy content-length/type because we may modify the response.
-  static final Set<String> HEADERS_TO_KEEP = Sets.newHashSet("last-modified", "date");
+  static final Set<String> RESPONSE_HEADERS_TO_KEEP =
+      Sets.newHashSet(
+          "last-modified",
+          "date",
+          "expires",
+          "content-location",
+          "etag",
+          "location",
+          "x-progress",
+          "x-request-id",
+          "x-correlation-id");
+
+  // The list of incoming header names to keep for forwarding request to the FHIR server; use lower
+  // case only.
+  // We should NOT copy Content-Length as this is automatically set by the RequestBuilder when
+  // setting content Entity; otherwise we will get a ClientProtocolException.
+  static final Set<String> REQUEST_HEADERS_TO_KEEP =
+      Sets.newHashSet(
+          "content-type",
+          "last-modified",
+          "etag",
+          "prefer",
+          "fhirVersion",
+          "if-match",
+          "if-none-match",
```

Comment:
Should we also include `if-none-exist`? In general, how can we make sure that these two sets are exhaustive?